[WFLY-10709] Support Eclipse MicroProfile OpenTracing

jpkrohling opened the pull request:
Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

cc @jmesnil , @pavolloffay, @rsvoboda 

This is a first WIP draft and is based on #11320 and https://github.com/jpkrohling/java-wildfly-tracing . 

JIRA: https://issues.jboss.org/browse/WFLY-10709
Proposal: https://github.com/wildfly/wildfly-proposals/pull/106

When deploying an application such as https://github.com/jpkrohling/java-wildfly-tester , a `curl` call like the one below results in a trace similar to the one in the image.

```bash
curl -X POST localhost:8080/java-wildfly-tester-1.0-SNAPSHOT/v1/order -v
```

![image](https://user-images.githubusercontent.com/13387/43151258-fdf8d6ca-8f6b-11e8-81e3-82b092d2cb21.png)

Please make sure your PR meets the following requirements:
- [x] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue(s)
- [x] Pull Request contains description of the issue(s)
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [x] Pull Request requires a change to the documentation
- [x] Documentation have been updated accordingly
- [x] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

jpkrohling committed:
[WFLY-10709] Support Eclipse MicroProfile OpenTracing

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

pavolloffay commented:
@jpkrohling trace on the picture does not use Microrofile-Opentraicng (smallrye) operation names. Are the operation names overridden or it's just an old picture? Also there should not be `/serialize` span if you want to pass TCK.

rsvoboda commented:
Juca, you are so quick.

Before proceeding with development, wf proposal https://github.com/wildfly/wildfly-proposals/pull/105 must be reviewed and agreed on, I still miss feedback from relevant QE and from Jeff.

Also PR should be created when feature is finished, reviewed and tested. Dev and QE are supposed to work on top of feature branch. Master has to have finalized stuff so it's "releasable anytime".

If you wanted to have our CI runs for your code, TeamCity supports custom runs against any repo and branch.

My understanding was that MP OT will be work for WildFly 15 and for WildFly 14 we will be focused on MP Config, MP Health and Logging enhancements for OpenShift.

jpkrohling commented:
> trace on the picture does not use Microrofile-Opentraicng (smallrye) operation names

My bad, I was using `opentracing-jaxrs`. I just switched to SmallRye's.

> I still miss feedback from relevant QE and from Jeff.

I'll ping Jeff, would you ping them today, thanks!

> Also PR should be created when feature is finished, reviewed and tested. Dev and QE are supposed to work on top of feature branch. Master has to have finalized stuff so it's "releasable anytime".

Understood. Should this feature branch located under my repository?

> My understanding was that MP OT will be work for WildFly 15 and for WildFly 14 we will be focused on MP Config, MP Health and Logging enhancements for OpenShift.

I thought this was for WF14, but might have misunderstood. The JIRA WFLY-10709 also seems to mention WF 14: 

> This requirement can be delivered as Technology Preview for WF 14.

The reason I'm trying to get this done quick is because I'm on PTO for two weeks starting this Friday. Looks like either we'll have to continue after my return or someone will have to take it over.

kabir commented:
This is ok to test

kabir commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,122 @@
+
+  private void addListeners(DeploymentUnit deploymentUnit) {
+    WarMetaData warMetaData = deploymentUnit.getAttachment(WarMetaData.ATTACHMENT_KEY);
+    JBossWebMetaData jbossWebMetaData = warMetaData.getMergedJBossWebMetaData();
```
I see some NPEs here in the test runs
```
Caused by: java.lang.NullPointerException
	at org.wildfly.extension.microprofile.opentracing.smallrye.TracingDeploymentProcessor.addListeners(TracingDeploymentProcessor.java:60)
	at org.wildfly.extension.microprofile.opentracing.smallrye.TracingDeploymentProcessor.deploy(TracingDeploymentProcessor.java:32)
	at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:144)
	... 8 more
```

kabir commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_OpenTracing_SmallRye.adoc:
```diff
@@ -0,0 +1,34 @@
+[source,ruby]
+----
+[standalone@localhost:9990 /]/extension=org.wildfly.extension.microprofile.opentracing-smallrye:add
+----
```
You also need to add the subsystem for the functionality to take effect

jpkrohling replied:
cc @jmesnil , as it probably applies to his MP-Config as well:

https://github.com/jmesnil/wildfly/blob/babea3b0f925fab044a3e9a04bfe30788f9b58c5/docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_Config_SmallRye.adoc

kabir commented on feature-pack/src/main/resources/modules/system/layers/base/org/eclipse/microprofile/opentracing/main/module.xml:
```diff
@@ -0,0 +1,22 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<module xmlns="urn:jboss:module:1.3" name="org.eclipse.microprofile.opentracing">
```
We should use the latest version 1.5

kabir commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -0,0 +1,57 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<module xmlns="urn:jboss:module:1.3" name="org.wildfly.extension.microprofile.opentracing-smallrye">
```
1.5

rhusar replied:
@kabir Actually, we are now up to 1.8 :-) https://github.com/jboss-modules/jboss-modules/blob/1.8.5.Final/src/main/resources/schema/module-1_8.xsd

kabir commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/DependencyProcessor.java:
```diff
@@ -0,0 +1,36 @@
+import org.jboss.modules.ModuleIdentifier;
+import org.jboss.modules.ModuleLoader;
+
+public class DependencyProcessor implements DeploymentUnitProcessor {
```
Nitpick: Other subsystems prefix the subsystem name, so something like OpenTracingDependencyProcessor

jpkrohling replied:
This is based on MP-Config, but I can certainly change it.

https://github.com/jmesnil/wildfly/blob/babea3b0f925fab044a3e9a04bfe30788f9b58c5/microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java#L40

kabir commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,122 @@
+
+  private void addJaxRsIntegration(DeploymentUnit deploymentUnit) {
+    WarMetaData warMetaData = deploymentUnit.getAttachment(WarMetaData.ATTACHMENT_KEY);
+    JBossWebMetaData jbossWebMetaData = warMetaData.getMergedJBossWebMetaData();
```
This can probably also be null

kabir commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingLogger.java:
```diff
@@ -0,0 +1,42 @@
+import org.jboss.logging.annotations.Message;
+import org.jboss.logging.annotations.MessageLogger;
+
+@MessageLogger(projectCode = "EMPTRAC", length = 4)
```
What is the 'E' for? :)
Apart from that I parse this to something like MicroProfile Trace/-ing

jpkrohling replied:
Eclipse, I believe. I based this on @jmesnil's MP-Config:

https://github.com/jmesnil/wildfly/blob/babea3b0f925fab044a3e9a04bfe30788f9b58c5/microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/MicroProfileConfigLogger.java#L37

bstansberry replied:
@jmesnil Do you still want to follow the project code convention at https://github.com/wildfly/wildfly/pull/11320/files#diff-93ee82755e25dc17584f3efd8b7fde11R37 now that the code is being moved into WIldFly itself?

I'm ok with it (and what Juraci's done here) if you do; I just wanted to double check in case it was something you'd have changed if you'd noticed.

jmesnil replied:
@bstansberry thanks for noticing i, I forgot about this.

We should change that to WFLYTRAC (and I'll do the same for the mp-config subsystem).
I used the EMPCONF code before the subsystem was brought into WildFly codebase. There is no good reason to diverge from the existing convention of WFLYXXX.

@jpkrohling could you update it please?

jpkrohling replied:
Done

kabir commented on pom.xml:
```diff
@@ -835,6 +845,12 @@
+            <dependency>
+                <groupId>${project.groupId}</groupId>
+                <artifactId>wildfly-microprofile-opentracing-smallrye</artifactId>
+                <version>${project.version}</version>
```
This and all other compile scope dependencies should have exclusions

jpkrohling replied:
May I ask you to confirm that this should happen *here*, at the root pom? It looks like the other dependencies listed here don't have any exclusions. The `feature-pack/pom.xml` do have them, though.

kabir replied:
@jamezp @stuartwdouglas @bstansberry WDYT?

When I looked quickly earlier I saw some dependencies having exclusions. A closer look shows it isn't totally consistent, so there are other dependencies which don't have exclusions.

jamezp replied:
I think we've gone about it a couple different ways. I'm personally of the feeling that third-party dependencies should exclude all dependencies and include any implicit required dependencies. However for dependencies that are a maven module of the project I think it's okay to not exclude all dependencies.

jpkrohling replied:
Could someone please confirm which one should I use here in the main pom? Looks like excluding everything is not a rule. In fact, most dependencies on this pom are *not* excluding everything.

I would not mind excluding everything, I just want to follow what's the common practice.

kabir reviewed (changes requested):
I've by no means done a complete review, but I have made a few comments

kabir reviewed (changes requested):
@jpkrohling It struck me that there should be some tests.

1) Every subsystem has a unit test to test that the parsing is ok. Look for BeanValidationSubsystemTestCase which is for a subsystem with a similarly simple/empty subsystem and copy/rename that to your subsystem unit tests.

2) There should be some integration tests (e.g. in testsuite/integration/basic or wherever else is appropriate). However, for all I know you might have some agreement with QE that they will deliver these. If the latter case, they should open a PR to your branch so they are included in this PR.

kabir commented:
Adding the 'hold' label after speaking to @rsvoboda. Ping me/ @jmesnil / @jamezp when it is ready

jpkrohling commented:
Status: the build is mostly passing, except for the very last module (Smoke tests), where one test is failing due to https://github.com/smallrye/smallrye-opentracing/issues/17 .

bstansberry commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,150 @@
+    if (null == serviceName || serviceName.isEmpty()) {
+      serviceName = deploymentUnit.getServiceName().getSimpleName();
+      TracingLogger.ROOT_LOGGER.serviceNameDerivedFromDeploymentUnit(serviceName);
+      System.setProperty("JAEGER_SERVICE_NAME", serviceName);
```
What if there are multiple deployments?

jpkrohling replied:
Great question. I'll change it so that we only *read* properties/env vars.

bstansberry commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDependencyProcessor.java:
```diff
@@ -0,0 +1,36 @@
+    ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
+    ModuleLoader moduleLoader = Module.getBootModuleLoader();
+
+    moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, EXTENSION_MODULE, false, true, true, false));
```
I think some sort of split up of the org.wildfly.extension.microprofile.opentracing-smallrye module is needed. We shouldn't be exposing subsystem modules to deployments. There are a couple places where we do and they're regrettable.

I commented on the MP Config PR the same way.

jpkrohling replied:
The first version of this PR had a very granular set of modules, but I ended up having CDI problems (beans registered in one module couldn't be seen/consumed by another module).

Would it be OK to have this as a JIRA to be completed before this gets promoted from tech-preview?

bstansberry commented:
Can you comment, preferably at the relevant point in the code, if not there then in the proposal, how all the Jaeger stuff gets brought in, configured etc. I don't mean the config details, e.g. that it reads env vars, but rather the hook that causes it all to happen. It's quite magical now. My guess is it's done as part of loading the CDI Extension?

Related to that is what of that is static, i.e. done once per VM?  Anything done once per VM is problematic is it cannot be updated via a server reload; it requires a full process restart.

jpkrohling commented:
> Can you comment, preferably at the relevant point in the code, if not there then in the proposal, how all the Jaeger stuff gets brought in, configured etc.

Right now, we are using the tracer resolver, which uses Java's ServiceLoader behind the scenes:

https://github.com/wildfly/wildfly/pull/11454/files#diff-968be69b1c607965089a660b4fdebf76R19

https://github.com/opentracing-contrib/java-tracerresolver/blob/353f318a883bc4f1675c7565e4d0a68c8af909b4/opentracing-tracerresolver/src/main/java/io/opentracing/contrib/tracerresolver/TracerResolver.java#L75

https://github.com/jaegertracing/jaeger-client-java/blob/f3e504992d19843245872fdd1c9f0006a3b70b57/jaeger-tracerresolver/src/main/java/io/jaegertracing/tracerresolver/internal/JaegerTracerResolver.java#L25

> Related to that is what of that is static, i.e. done once per VM? Anything done once per VM is problematic is it cannot be updated via a server reload; it requires a full process restart.

Great point. I think the only place where it could be a problem is with the [GlobalTracer](https://github.com/opentracing/opentracing-java/blob/master/opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java) . 

cc @pavolloffay, @objectiser : can you think of any other places where we might get bitten by this?

objectiser commented:
@jpkrohling no, global tracer is the only one - and that is scoped to the deployment classloader so assume will be replaced on server reload.

jpkrohling commented:
> Right now, we are using the tracer resolver, which uses Java's ServiceLoader behind the scenes:

This is not true anymore, as I'm having Jaeger-specific code now inside the `TracerInitializer`, so that we don't need to write a system property with the service name (https://github.com/wildfly/wildfly/pull/11454#discussion_r207976139).

pavolloffay commented:
Is the `GlobalTracer` required? Why not use only CDI? MP-OpenTracing defines that tracer should be injectable.

jpkrohling commented:
@pavolloffay wrote:
> Is the GlobalTracer required? Why not use only CDI? MP-OpenTracing defines that tracer should be injectable.

You are right. If I create the SpanFinishingFilter on the ServletContextListener, I don't need the GlobalTracer. But the underlying application might still expect to find the tracer there, as it's the usual idiom in OT.

jpkrohling commented:
Status (cc @jmesnil, @rsvoboda, @bstansberry):

I added a simple basic test to make sure that the tracer can be injected into the application. I also addressed some concerns from reviewers of this PR. The only thing that is still open is about the granularity of the modules. Also on my list is the inclusion of more tests, to assess other aspects of this work, like, to make sure spans are created when JAX-RS requests are made.

I'd like to understand first, though, what would be required to get this PR merged as Technology Preview and prioritize those tasks.

bstansberry commented:
@jpkrohling Thanks for the pointer to the ServletContextInitializer; I missed that. So this opens a client socket per webapp. I guess that's ok, and in any case it's an internal detail.

Here's what I think are the most important things:

1) I assume the MP OT TCK covers a lot of scenarios that fit typical WildFly Swarm use cases. So since that's more covered, some testing of WildFly oddities is good. Specifically we support ears, we support multiple deployments and it's common for people to reload the server. So a test of an ear that contains 2 wars (which therefore kind of covers the multiple deployment case) and also does a reload would be good.  @yersan can help you out with the reload part if you need that.

2) We should clarify the GlobalTracer thing.  It is not scoped to the deployment classloader in WildFly; the class is part of a static module so there's only going to be one instance per server. Looking at it, it seems it expects someone to call GlobalTracer.register. It should not be any deployment code that succeeds in doing that. But the class is exposed to a deployment so to prevent a deployment setting the tracer the subsystem should. But I don't know the implications of that, particularly if a reload happens. I don't know what kind of state a Tracer keeps, e.g. if keeping one in GlobalTracer across reloads might leak a classloader.

3) About the module thing we can defer creating another module but the code should be reorganized so the stuff that needs to be visible to the deployment is in a different package from the subsystem stuff. That's better organization regardless and let's us look into simple things like adding a module that simply has a module.xml that depends on the current one and exports 1 package.

objectiser commented:
For the GlobalTracer issue, could we register a `Tracer` proxy in the global tracer, that when called will try to use a tracer injected in the local deployment context? Hopefully the `GlobalTracer` will not be used, but in case the deployment does, it would be good if it just deferred to CDI.

jpkrohling commented:
> Looking at it, it seems it expects someone to call GlobalTracer.register. It should not be any deployment code that succeeds in doing that.

I'll need some help with JBoss Modules here to see if we are doing the right thing... We cannot remove a dependency on the GlobalTracer, as some instrumentation libraries (JAX-RS, for instance) has references to it, although that code shouldn't be reached.

> About the module thing we can defer creating another module but the code should be reorganized so the stuff that needs to be visible to the deployment is in a different package from the subsystem stuff.

Done: I moved the initialized, producer and CDI extension to a `deployment` package.

yersan commented:
@jpkrohling FYI. I don't know the details, but, if you need to do any reload, there is a handy class for such purposes: https://github.com/wildfly/wildfly/blob/master/testsuite/shared/src/main/java/org/jboss/as/test/shared/ServerReload.java

rsvoboda commented on testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/microprofile/opentracing/ResourceWithCustomOperationNameBeanTestCase.java:
```diff
@@ -54,7 +55,13 @@
     }
 
     @Test
+    @Ignore
```
General comment: @Ignore("some reasoning here ...") is helpful as TeamCity can show these messages and we have testsuite analyzer which looks into this when we are looking into details about ignored or assumed tests

jpkrohling commented:
I believe this is now ready for a code review. cc @bstansberry, @jmesnil

@kabir: do you think the `hold` label could be removed as well?

jamezp commented on feature-pack/pom.xml:
```diff
@@ -2802,6 +2807,140 @@
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
```
All these new dependencies need to be added to the `feature-pack/src/license/full-feature-pack-licenses.xml`.

jamezp commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -0,0 +1,58 @@
+        <module name="javaee.api"/>
+        <module name="javax.enterprise.api" />
+
+        <module name="org.slf4j" export="true"/>
```
This should probably not be exported.

jamezp commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -0,0 +1,58 @@
+        <artifact name="${io.jaegertracing:jaeger-thrift}"/>
+
+        <artifact name="${com.google.code.gson:gson}"/>
+    </resources>
```
It's unusual to have this many artifacts in the same module. Generally these should be modules on their own that `org.wildfly.extension.microprofile.opentracing-smallrye` depends on.

jpkrohling replied:
+1 . Now that we have everything working and TCK passing, I was able to move things into their own modules and got it almost working (works on some tests already). The current problem is that I have two classes under the same package, but different modules, due to https://github.com/opentracing-contrib/java-tracerresolver/issues/31 . I have this worked saved in a different branch and will resume working on it if we can get a fix/release of ot-tracerresolver in time.

jamezp commented on microprofile/opentracing-smallrye/pom.xml:
```diff
@@ -0,0 +1,106 @@
+        <dependency>
+            <groupId>org.wildfly.core</groupId>
+            <artifactId>wildfly-subsystem-test</artifactId>
+            <type>pom</type>
```
This should be `<scope>test</scope>`. It also seems kind of odd to have the `<type>pom</type>` as well. Generally the dependency should be:
```xml
<dependency>
    <groupId>org.wildfly.core</groupId>
    <artifactId>wildfly-subsystem-test-framework</artifactId>
    <scope>test</scope>
</dependency>
```

jpkrohling replied:
I'm almost certain I copied this from the MP-Config, so, this source might be wrong as well:

https://github.com/wildfly/wildfly/blob/master/microprofile/config-smallrye/pom.xml#L88

cc @jmesnil

jamezp replied:
At a minimum it should at least be `<scope>test</scope>`.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/io/opentracing/contrib/tracerresolver/NaiveTracerResolver.java:
```diff
@@ -0,0 +1,40 @@
+package io.opentracing.contrib.tracerresolver;
```
This should probably have a copyright header.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemDefinition.java:
```diff
@@ -0,0 +1,30 @@
+import org.jboss.as.controller.registry.ManagementResourceRegistration;
+
+public class SubsystemDefinition extends PersistentResourceDefinition {
+    static AttributeDefinition[] ATTRIBUTES = {};
```
This should probably be final as I don't see the array being added to anywhere. Or really it could just be deleted.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemDefinition.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public Collection<AttributeDefinition> getAttributes() {
+        return Arrays.asList(ATTRIBUTES);
```
Since it's an empty array it could just return `Collections.emptyArray()`.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemDefinition.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public void registerOperations(ManagementResourceRegistration resourceRegistration) {
+        super.registerOperations(resourceRegistration);
+    }
```
This is a bit nitpicky, but there probably isn't really a need to override this.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemExtension.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public void initialize(ExtensionContext context) {
+        final SubsystemRegistration subsystem = context.registerSubsystem(SUBSYSTEM_NAME, CURRENT_MODEL_VERSION);
+        subsystem.registerXMLElementWriter(() -> SubsytemParser_1_0.INSTANCE);
```
Since the instance is static we should just use the non-Supplier method. `subsystem.registerXMLElementWriter(SubsytemParser_1_0.INSTANCE);`

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemExtension.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void initializeParsers(ExtensionParsingContext context) {
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, SubsytemParser_1_0.NAMESPACE, () -> SubsytemParser_1_0.INSTANCE);
```
Same with this one:
`context.setSubsystemXmlMapping(SUBSYSTEM_NAME, SubsytemParser_1_0.NAMESPACE, SubsytemParser_1_0.INSTANCE);`

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/SubsystemRemove.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) {
+        log.info("Removing tracing subsystem");
```
This should be a debug or trace message and should use the `TracingLogger.ROOT_LOGGER.debug("Removing tracing subsystem")`. It also seems a little odd to me that a remove handler doesn't really do anything. This likely should just use the `ReloadRequiredRemoveStepHandler.INSTANCE` for the remove handler making this obsolete.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDependencyProcessor.java:
```diff
@@ -0,0 +1,38 @@
+  public static final ModuleIdentifier MP_OT_MODULE = ModuleIdentifier.create("org.eclipse.microprofile.opentracing");
+
+  public static final Phase PHASE = Phase.DEPENDENCIES;
+  public static final int PRIORITY = 0x4000;
```
@jmesnil and I discussed this a bit on https://github.com/wildfly/wildfly/pull/11320. We may need a follow JIRA depending on what we decide to do here.

jpkrohling replied:
I changed this to be based on what was merged for MP-Config

jamezp replied:
Ah. I missed that. It won't be allowed to have the same phase and priority. @jmesnil do you have an opinion on this? Mainly what the priority should be.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,124 @@
+
+public class TracingDeploymentProcessor implements DeploymentUnitProcessor {
+  public static final Phase PHASE = Phase.POST_MODULE;
+  public static final int PRIORITY = 0x4000;
```
Same comment about the Phase and the discussion on https://github.com/wildfly/wildfly/pull/11320

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,124 @@
+
+    if (!WeldDeploymentMarker.isPartOfWeldDeployment(deploymentUnit)) {
+      // SmallRye JAX-RS requires CDI. Without CDI, there's no integration needed
+      TracingLogger.ROOT_LOGGER.noCdiDeployment();
```
Should this be logging an error or throwing an exception? If it requires CDI it seems the deployment should fail.

jpkrohling replied:
Good point. As tracing is usually not something that is a requirement for a business code to run (like logging), we tend to just warn users instead of failing their applications.

@objectiser , what do you think? Should we just warn, or should we fail the deployment?

objectiser replied:
My preference would be just to warn at this stage.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  private String getServiceName(DeploymentUnit deploymentUnit) {
+    String serviceName = System.getProperty("JAEGER_SERVICE_NAME");
```
This should use `WildFlySecurityManager.getPropertyPrivileged()`. That said I'm not sure I understand this. Controlling a service name depending on a system property or environment variable seems risky.

jpkrohling replied:
Would you mind enumerating the risks? I'm not sure I fully understand them.

The idea is to keep the user's expectation regarding this env var/sysprop: Jaeger's service name is the only required configuration entry for Jaeger to work and it's usually set via env var/sysprop. *Not* supporting this way of configuring would make WildFly an exception and will probably confuse users.

We'll add a way to control this via WildFly's management model, but we'll likely keep supporting the configuration via env var.

jamezp replied:
My apologies. I was thinking of an MSC `ServiceName`. This should use the `WildFlySecurityManager` though so the deployment will work with a security manager installed.

jpkrohling replied:
Alright, I just changed it. I believe all your review items are now addressed.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingDeploymentProcessor.java:
```diff
@@ -0,0 +1,124 @@
+  private String getServiceName(DeploymentUnit deploymentUnit) {
+    String serviceName = System.getProperty("JAEGER_SERVICE_NAME");
+    if (null == serviceName || serviceName.isEmpty()) {
+      serviceName = System.getenv("JAEGER_SERVICE_NAME");
```
This should use `WildFlySecurityManager.getEnvPropertyPrivileged()`. Same comment about the risk as well.

jamezp commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/extension/microprofile/opentracing/smallrye/TracingLogger.java:
```diff
@@ -0,0 +1,73 @@
+
+  @LogMessage(level = DEBUG)
+  @Message(id = 2, value = "MicroProfile OpenTracing Subsystem is processing deployment")
+  void processingDeployment();
```
I see no need to remove these, but just as an FYI we only require INFO and higher to be internationalized.

jpkrohling replied:
Those were INFO in an earlier iteration, but I downgraded them to DEBUG, as I found the log too verbose for the subsystem.

jamezp reviewed (changes requested):
There are some TODO's in here that make me a little nervous if this is going into WF 14 as the deadline for features is very soon. Someone with more knowledge of this feature should probably have a look at them.

There is also a lot of missing copyright headers. From a personal perspective I like them for the consistency, but someone else is welcome to make the call on whether they are required or not :)

jpkrohling commented:
> There are some TODO's in here that make me a little nervous if this is going into WF 14 as the deadline for features is very soon

Two of the TODOs are about bugs found during the development of this PR. If we get a release from those components by the time this PR gets merged, those TODOs will be removed. Otherwise, they are just reminders that something has to be done.

One is about a question, but I decided to remove, as there's a test that hits that condition already.

The other is about a decision we made already, so, I just removed.

I updated this PR with fixes to most of the comments by @jamezp . The current state passes the TCK as well. Once this is deemed ready to be merged, I can squash the commits, but I'll leave with individual commits for now, to make it easier to get a re-review.

mjurc committed:
[WFLY-10709] added simple integration tests

jpkrohling commented:
All TODO's are now addressed: I released a new version of Jaeger Client Java and got a new release of Tracer Resolver as well, so that we are now in a good state to get this merged.

I tested this PR at its current state and it passes the TCK.

jamezp commented on microprofile/opentracing-extension/src/test/resources/org/wildfly/extension/microprofile/opentracing/subsystem_1_0.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
This actually needs to be removed. The test automatically adds this so it ends up in the file twice which is why the unit test is failing.

jamezp reviewed (changes requested):
One more minor thing. I noticed that some of the source is using 2 spaces instead of 4 for formatting. We should keep it at 4 for consistency with the rest of the code base :)

jpkrohling commented:
> I noticed that some of the source is using 2 spaces instead of 4 for formatting. 

Sorry, my bad! Should be fixed in the latest commit. 

As usual, the current state of the PR is confirmed to pass the TCK.

kabir commented:
@jpkrohling I think the mixed domain test failures in the linux and windows jobs are because you need to do something similar to https://github.com/wildfly/wildfly/pull/11503/files#diff-a638e7e043f2b0642d831a9f9069c59eR64 to remove the new extension and subsystem. 

These tests are not run by default as there is some setup involved. If you want to run them locally, you need to download EAP 6.4.0, 7.0.0 and 7.1.0 and put the zips into a folder (e.g `/some/where`). Then to run these tests you execute: `mvn clean install -pl testsuite/mixed-domain -DallTests -Djboss.test.mixed.domain.dir=/some/where`

jpkrohling commented:
Looks like the tests are passing now, thanks! The only one still failing is "Linux JDK10 (Pull Request) - merge", but looks like it's failing due to some unrelated SSL tests.

Is there anything still missing, or is this PR ready to get merged?

![image](https://user-images.githubusercontent.com/13387/44087178-3158e132-9fbf-11e8-8fb3-22de39e638df.png)

jstourac commented:
Test-dev shall be done. There are two failing tests:
```
[ERROR] Failures: 
[ERROR]   EarOpenTracingTestCase.testEarServicesUseDifferentTracers:79->testHttpInvokation:98 Service one and service two tracer instance hash is same - 1091667486. Actual: 1091667486
[ERROR]   EarOpenTracingTestCase.testEarServicesUseDifferentTracersAfterReload:85->testHttpInvokation:98 Service one and service two tracer instance hash is same - 1091667486. Actual: 1091667486
```
These tests are currently marked to be ignored. There is a tracker jira to resolve root issue https://issues.jboss.org/browse/WFWIP-104. The issue shall be fixed after the feature is merged into master.

We are okay to merge this.

bstansberry commented:
I filed https://issues.jboss.org/browse/WFWIP-108, which we need to fix ASAP but doesn't need to block merging this for Beta2.

bstansberry commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/MicroProfile_OpenTracing_SmallRye.adoc:
```diff
@@ -0,0 +1,73 @@
+WARNING: by default, the Jaeger Client Java has a probabilistic sampling strategy, set to 0.001, meaning that only
+approximately one in one thousand traces will be sampled. To sample every request, set the environment variable
+`JAEGER_SAMPLER_TYPE` to `const` and `JAEGER_SAMPLER_PARAM` to `1`.
+
```
Please add a paragraph something like this:

WARNING: the supported configuration options for this Tech Preview feature, particularly those related to configuring the Jaeger Java Client tracer, may change in incompatible ways in a future release.

bstansberry commented:
I made a couple doc comments that I'd like in for .Final but again those should not block merging this.

bstansberry commented on testsuite/integration/basic/src/test/config/arq/arquillian.xml:
```diff
@@ -7,7 +7,7 @@
         <configuration>
             <property name="jbossHome">${basedir}/target/wildfly</property>
-            <property name="javaVmArguments">${server.jvm.args}</property>
+            <property name="javaVmArguments">${server.jvm.args} -agentlib:jdwp=transport=dt_socket,address=8787,server=y,suspend=n</property>
```
This looks like development cruft. We use maven property surefire.jpda.args to control this kind of thing. That's already pretty complex so we shouldn't add other one-off things.

jpkrohling replied:
Oops, sorry! Will send a PR right after this gets merged.

bstansberry reviewed (approved):
This looks ok for Beta2.

I commented re docs, jboss-api=private and a jdwp config setup in an arquillian.xml file but all those can be corrected after this is merged.
